src/device/ping-device.cpp

```diff
@@ -0,0 +1,93 @@
+#include <ping-device.h>
+
+#include <ping-message-common.h>
+#include <ping-message.h>
+#include <ping-time.h>
+
+bool PingDevice::initialize() {
+  if (!request(CommonId::PROTOCOL_VERSION)) {
+    return false;
+  }
+
+  if (!request(CommonId::DEVICE_INFORMATION)) {
+    return false;
+  }
+
+  return true;
+}
+
+ping_message* PingDevice::read() {
+  char b;
+  int result = _port.read(&b, 1);
+  if (result > 0) {
+    if (_parser.parseByte((char)b) == PingParser::NEW_MESSAGE) {
+      return &_parser.rxMessage;
+    }
+  }
+  return nullptr;
+}
+
+ping_message* PingDevice::request(uint16_t id, int timeoutMs) {
+  common_general_request msg;
+  msg.set_requested_id(id);
+  writeMessage(msg);
+  return waitMessage(id, timeoutMs);
+}
+
+ping_message* PingDevice::waitMessage(uint16_t id, int timeoutMs) {
+  int tstart = PingTime::timeMs();
+  while (PingTime::timeMs() < tstart + timeoutMs) {
+    ping_message* msg = read();
+
+    if (!msg) {
+      continue;
+    }
+
+    _handleMessage(msg);
+
+    if (msg->message_id() == CommonId::NACK) {
+      return nullptr;
```

Comment:
Why are you returning NACK for a request ? The library do not handle NACKs ?

Reply:
This is not request, it is `waitMessage(uint16_t id). It returns null unless the message id matched. I will see if I can make this more useful with nack.

Reply:
you are right it affects request, the point of request/waitmessage is to return only the message id you ask for, else nullptr.

I am changing it to return the message you ask for, nack if the nack id matches the messaage you ask for, else nullptr.

Reply:
done, the nack message is returned instead of nullptr.